### ISSUE TITLE
`function.__doc__` should type-narrow to `str`.

### ISSUE THREAD

**Issue by aspizu** [microsoft/pyright#4840]
```py
def foo():
    """docstring"""
    ...

foo.__doc__ # Type is str | None, but it should be str.
```

**Comment by erictraut** [comment]
The type for `__doc__` comes from the class definition for `object`, and it's defined as `str | None`. It's not special-cased in the type checker. This is intentional. If you want to narrow it yourself, you can add an assert.

```python
assert foo.__doc__ is not None
```

**Comment by Azureblade3808** [comment]
From what I've learned, please DON'T do this --
```python
assert foo.__doc__ is not None
```

When your code gets run in optimized mode level 2, all doc-strings and assertions will get discarded. So there is a chance this assertion should really fail, but meanwhile it would never function at all. Therefore, you may get confused why your code no longer works.

Try this --
```python
# test.py

def foo():
    """docstring"""
    ...

assert foo.__doc__ is not None
print(len(foo.__doc__))
```
and run it by --
```
python -OO test.py
```
As a result, you will get a `TypeError` instead of an `AssertionError`.
